docs/contrib: expand the contribution guide (#854)
amaranth-lang/amaranth: docs/contrib.rst

> The Amaranth project is a collective work of many people collaborating over the years, and it would not be the same without everyone's unique perspectives and contributions. We're glad that you are considering joining us! This page will guide you through some of the ways to contribute.

Reviewer: ```suggestion
The Amaranth project is the collective work of many people collaborating over the years, and it would not be the same without everyone's unique perspectives and contributions. We're glad that you are considering joining us! This page will guide you through some of the ways to contribute to the project.
```

Filing problem reports
======================

We would like Amaranth to be a best-in-class design tool, and hearing about issues people encounter while using it is crucial for improving it. While we do care a lot about correctness of the results, we care about the experience of using the tool just as much. Amaranth is meant to be a tool that is comfortable to use: with fewer sharp edges (no matter how much technological appeal they might have) and more signs and guardrails.

> Please `report <issues>`_ any problems you encounter using Amaranth. To go beyond that: **If, while you are using Amaranth, you see an error message that is hard to understand or is misleading, please report it as a bug. Even (especially!) if you were doing something wrong.**

Reviewer: ```suggestion
Please `report <issues>`_ any problems you encounter using Amaranth. To go beyond that: **If, while you are using Amaranth, you see an error message that is hard to understand or is misleading, please report it as a bug. Even (especially!) if you think you did something wrong.**
```

.. _issues: https://github.com/amaranth-lang/amaranth

When filing problem reports, please include the following information:

> * The exact version of Amaranth which you can find it out by running ``python -c "import amaranth; print(amaranth.__version__)"``;

Reviewer: ```suggestion
* The exact version of Amaranth, which you can find by running ``python -c "import amaranth; print(amaranth.__version__)"``;
```

* A complete, self-contained, and minimal program that demonstrates the problem you are reporting (if minimizing it is not feasible, include the exact sequence of steps that reproduces the problem);

> * What did you expect to happen, and what actually happened (where possible, including a verbatim copy of the log file or the terminal output);

Reviewer: ```suggestion
* What you expected to happen, and what actually happened (where possible, including a verbatim copy of the log file or the terminal output);
```

> * For usability issues: your reasoning in filing the report (i.e. why did you expect a different behavior).

Reviewer: ```suggestion
* For usability issues: your reason for filing the report (i.e. why did you expect a different behavior).
```

Preparing the environment
-------------------------

> The Amaranth codebase uses the PDM_ package and dependency manager to structure the development workflow. Please `install PDM`_ first and make sure you have downloaded the latest changes to the sources. Once you have done so, run:

Reviewer: ```suggestion
The Amaranth codebase uses the PDM_ package and dependency manager to structure the development workflow. Please `install PDM`_ first and make sure you have downloaded the latest changes to the source files. Once you have done so, run:
```

.. todo::

   This section is unfinished

At the moment there is no formal style guide for source code. We ask that you do your best effort to keep the code that you add or modify similar in style as well as in spirit to the code surrounding it, and we may ask you to change it during review. When in doubt, submit your code as-is.

Weekly meetings
===============

> Every Monday at 17:00 UTC on our IRC channel `#amaranth-lang at libera.chat <irc>`_ or Matrix channel `#amaranth-lang@matrix.org <matrix>`_ (the channels are bridged together: the same messages appear on both), Amaranth maintainers meet with users and contributors to discuss newly submitted Requests for Comments and any other issues that warrant broad attention. These public meetings are the primary avenue of decision making.

Reviewer: ```suggestion
Every Monday at 17:00 UTC on our IRC channel `#amaranth-lang at libera.chat <irc>`_ or Matrix channel `#amaranth-lang@matrix.org <matrix>`_ (the channels are bridged together: the same messages appear on both), Amaranth maintainers meet with users and contributors to discuss newly-submitted Requests for Comment and any other issues that warrant broad attention. These public meetings are the primary avenue of decision making.
```

Author: The singular is "Request for Comments"--wouldn't the plural be "Requests for Comments"?

Reviewer: Ah, I see. Then, yes, you're right. Often it's Request for Comment, so that's what tripped me up.

> Every Tuesday at 17:00 UTC on our IRC channel `#amaranth-lang at libera.chat <irc>`_ or Matrix channel `#amaranth-lang@matrix.org <matrix>`_ (the channels are bridged together: the same messages appear on both), Amaranth maintainers meet with users and contributors to discuss newly submitted Requests for Comments and any other issues that warrant broad attention. These public meetings are the primary avenue of decision making.

Reviewer 2: Should this say Monday?

Author: Yes. Oops...
